packages/pwa-buildpack/lib/WebpackTools/configureWebpack.js

```diff
@@ -0,0 +1,342 @@
+const RootComponentsPlugin = require('./plugins/RootComponentsPlugin');
+const ServiceWorkerPlugin = require('./plugins/ServiceWorkerPlugin');
+const UpwardDevServerPlugin = require('./plugins/UpwardDevServerPlugin');
+const UpwardIncludePlugin = require('./plugins/UpwardIncludePlugin');
+const PWADevServer = require('./PWADevServer');
+const MagentoResolver = require('./MagentoResolver');
+
+const prettyLogger = require('../util/pretty-logger');
+
+/**
+ * We need a root directory for the app in order to build all paths relative to
+ * that app root. It's not safe to use process.cwd() here because that depends
+ * on what directory Node is run in. The project root should be the dir of the
+ * webpack.config.js which called this function.
+ *
+ * There is no safe way to get the path of this function's caller, so instead we
+ * expect that the webpack.config.js will do:
+ *
+ *     configureWebpack(__dirname);
+ */
+async function validateRoot(appRoot) {
+    if (!appRoot) {
+        throw new Error(
+            'Must provide the root directory of the PWA to as the first parameter to `configureWebpack()`. In webpack.config.js, the recommended code is `configureWebpack(__dirname)`.'
+        );
+    }
+    // If root doesn't exist, an ENOENT will throw here and log to stderr.
+    const dirStat = await stat(appRoot);
+    if (!dirStat.isDirectory()) {
+        throw new Error(
+            `Provided application root "${appRoot}" is not a directory.`
+        );
+    }
+}
+
+async function checkForBabelConfig(appRoot) {
+    try {
+        await stat(path.resolve(appRoot, 'babel.config.js'));
+        return true;
+    } catch (e) {
+        return false;
+    }
+}
+
+function getMode(cliEnv = {}, projectConfig) {
+    if (cliEnv.mode) {
+        return cliEnv.mode;
+    }
+    if (projectConfig.isProd) {
+        return 'production';
+    }
+    return 'development';
+}
+
+async function configureWebpack({ context, vendor = [], special = {}, env }) {
+    await validateRoot(context);
+
+    const babelConfigPresent = await checkForBabelConfig(context);
+
+    let projectConfig = loadEnvironment(context);
+
+    const paths = {
+        src: path.resolve(context, 'src'),
+        output: path.resolve(context, 'dist')
+    };
+
+    const features = await Promise.all(
+        Object.entries(special).map(async ([packageName, flags]) => [
+            await pkgDir(path.dirname(require.resolve(packageName))),
+            flags
+        ])
+    );
+
+    const hasFlag = flag =>
+        features.reduce(
+            (hasIt, [packagePath, flags]) =>
+                flags[flag] ? [...hasIt, packagePath] : hasIt,
+            []
+        );
+
+    let mode = getMode(env, projectConfig);
+
+    const { debugBundles } = projectConfig.section('developer');
+
+    if (debugBundles) {
+        prettyLogger.warn(
+            'Debugging bundles. Large but readable files will be generated.'
+        );
+        projectConfig = loadEnvironment(
```

Comment:
Is it necessary to reload the environment to override a potential `NODE_ENV` value? Why not:

```js
projectConfig.env = Object.assign({}, projectConfig.env, {
  NODE_ENV: 'development'
});
```

Reply:
I'm trying to treat `Configuration` objects as immutable, so assigning to `projectConfig.env` felt wrong. Recreating and revalidating the object felt more respectful of the object's boundaries.

Reply:
Still, can you just pass the desired value in to the initial call?